src/main/java/com/google/devtools/build/lib/syntax/MethodDescriptor.java

```diff
@@ -59,19 +62,157 @@ private MethodDescriptor(
       boolean allowReturnNones,
       boolean useStarlarkThread,
       boolean useStarlarkSemantics) {
+    boolean isStringModule = method.getDeclaringClass() == StringModule.class;
+
     this.method = method;
     this.annotation = annotation;
     this.name = name;
     this.doc = doc;
     this.documented = documented;
     this.structField = structField;
-    this.parameters = parameters;
+    this.parameters =
+        !isStringModule ? parameters : Arrays.copyOfRange(parameters, 1, parameters.length);
     this.extraPositionals = extraPositionals;
     this.extraKeywords = extraKeywords;
     this.selfCall = selfCall;
     this.allowReturnNones = allowReturnNones;
     this.useStarlarkThread = useStarlarkThread;
     this.useStarlarkSemantics = useStarlarkSemantics;
+    try {
+      MethodHandle mh = MethodHandles.lookup().unreflect(method);
+
+      if (isStringModule) {
+        // String methods get the string as an extra argument
+        // because their true receiver is StringModule.INSTANCE.
+        mh = MethodHandles.insertArguments(mh, 0, StringModule.INSTANCE);
+      }
+
+      if (!useStarlarkThread) {
+        mh = MethodHandles.dropArguments(mh, mh.type().parameterCount(), StarlarkThread.class);
+      }
+
+      mh = MethodHandleUtil.moveArgument(mh, mh.type().parameterCount() - 1, 1);
+
+      // mh: (self, StarlarkThread, args...) -> ?
+
+      mh = mh.asSpreader(Object[].class, mh.type().parameterCount() - 2);
+
+      // mh: (self, StarlarkThread, args) -> ?
+
+      if (mh.type().returnType() == void.class) {
+        mh =
+            MethodHandles.filterReturnValue(
+                mh, MethodHandles.constant(NoneType.class, Starlark.NONE));
+        mh = MethodHandles.dropArguments(mh, mh.type().parameterCount(), Mutability.class);
+      } else if (Starlark.validValueClass(mh.type().returnType())) {
+        mh = handleNullReturn(mh, allowReturnNones, name);
+        mh = MethodHandles.dropArguments(mh, mh.type().parameterCount(), Mutability.class);
+      } else {
+        mh = handleNullReturn(mh, allowReturnNones, name);
+
+        MethodHandle fromJava =
+            MethodHandles.explicitCastArguments(
+                FROM_JAVA,
+                MethodType.methodType(Object.class, mh.type().returnType(), Mutability.class));
+        mh = MethodHandles.collectArguments(fromJava, 0, mh);
+      }
+
+      mh = MethodHandleUtil.moveArgument(mh, 3, 2);
+
+      Class<?> expectedP0 = isStringModule ? String.class : method.getDeclaringClass();
+      // Self-check: better fail earlier.
+      Preconditions.checkState(mh.type().parameterCount() == 4, "%s", mh.type());
```

Comment:
Clearer as:

if (cond || cond || cond || cond || cond) {
    throw new IllegalStateException(a single well-crafted error message)
}

Reply:
I dropped the assertion. It was useful mostly for development, the asserion is still here in `explicitlyCastArguments`.